src/nvim/ui.h

```diff
@@ -22,6 +23,7 @@ struct ui_t {
   void (*clear)(UI *ui);
   void (*eol_clear)(UI *ui);
   void (*cursor_goto)(UI *ui, int row, int col);
+  void (*cursor_style_set)(UI *ui, Dictionary args); // or a Dictionary ?
```

Comment:
nvim-qt never sees any of this, it communicates via msgpack-RPC types. That's where the `api/ui.c` "intermediate step" comes into play.

Only in-process UIs would be affected by the choice of type here. Currently VimR is the only example of that. The `private/` namespace is for the theoretical future where we have a "pubic ABI": `api/*` excluding `api/private/*`. 

But in the particular case of `api/private/defs.h`, it's included by `api/buffer.h` and friends:

     #include "nvim/api/private/defs.h"

So Dictionary, Buffer, and the other API C types are (intentionally) exposed in the public ABI. The purpose of the `private/` namespace is to hide the implementation, not the types.